### PR TITLE
[maistra-2.6] Rebase gateway controller

### PR DESCRIPTION
**Pull request by eoinfennessy**

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/controller.go
Comment (eoinfennessy) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:285]: Manually moved from `crdclient/cache_handler.go`

File: pilot/pkg/config/kube/crdclient/client.go
Comment (eoinfennessy) [inline pilot/pkg/config/kube/crdclient/client.go:359]: Moved from crdclient/cache_handler.go

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go
Comment (jewertow) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go]: Don't we also need the following changes?
```
diff --git a/pilot/pkg/config/kube/gateway/deploymentcontroller.go b/pilot/pkg/config/kube/gateway/deploymentcontroller.go
index da910e9153..e24e501fd0 100644
--- a/pilot/pkg/config/kube/gateway/deploymentcontroller.go
+++ b/pilot/pkg/config/kube/gateway/deploymentcontroller.go
@@ -122,7 +122,7 @@ var builtinClasses = getBuiltinClasses()
 
 func getBuiltinClasses() map[gateway.ObjectName]gateway.GatewayController {
        res := map[gateway.ObjectName]gateway.GatewayController{
-               defaultClassName:                 constants.ManagedGatewayController,
+               defaultClassName:                 controllerName,
                constants.RemoteGatewayClassName: constants.UnmanagedGatewayController,
        }
        if features.EnableAmbientControllers {
```
```
diff --git a/pilot/pkg/config/kube/gateway/deploymentcontroller.go b/pilot/pkg/config/kube/gateway/deploymentcontroller.go
index e24e501fd0..5c86f6941f 100644
--- a/pilot/pkg/config/kube/gateway/deploymentcontroller.go
+++ b/pilot/pkg/config/kube/gateway/deploymentcontroller.go
@@ -122,7 +122,7 @@ var builtinClasses = getBuiltinClasses()
 
 func getBuiltinClasses() map[gateway.ObjectName]gateway.GatewayController {
        res := map[gateway.ObjectName]gateway.GatewayController{
-               defaultClassName:                 controllerName,
+               defaultClassName:                 constants.ManagedGatewayController,
                constants.RemoteGatewayClassName: constants.UnmanagedGatewayController,
        }
        if features.EnableAmbientControllers {
@@ -133,8 +133,8 @@ func getBuiltinClasses() map[gateway.ObjectName]gateway.GatewayController {
 
 func getClassInfos() map[gateway.GatewayController]classInfo {
        m := map[gateway.GatewayController]classInfo{
-               constants.ManagedGatewayController: {
-                       controller:         constants.ManagedGatewayController,
+               controllerName: {
+                       controller:         controllerName,
                        description:        "The default Istio GatewayClass",
                        templates:          "kube-gateway",
                        defaultServiceType: corev1.ServiceTypeLoadBalancer,
```
If I correctly understand the reconciliation logic (lines 254-283), custom gateway classes will never match the controller name set in `FieldManager`:
```
// Reconcile takes in the name of a Gateway and ensures the cluster is in the desired state
func (d *DeploymentController) Reconcile(req types.NamespacedName) error {
	log := log.WithLabels("gateway", req)

	gw := d.gateways.Get(req.Name, req.Namespace)
	if gw == nil {
		log.Debugf("gateway no longer exists")
		// we'll ignore not-found errors, since they can't be fixed by an immediate
		// requeue (we'll need to wait for a new notification), and we can get them
		// on deleted requests.
		return nil
	}

	var controller gateway.GatewayController
	var gc *gateway.GatewayClass
	if d.gatewayClasses != nil {
		gc = d.gatewayClasses.Get(string(gw.Spec.GatewayClassName), "")
	}
	if gc != nil {
		controller = gc.Spec.ControllerName
	} else {
		if builtin, f := builtinClasses[gw.Spec.GatewayClassName]; f {
			controller = builtin
		}
	}
	ci, f := classInfos[controller]
	if !f {
		log.Debugf("skipping unknown controller %q", controller)
		return nil
	}
```

**Review by jewertow — COMMENTED** [review]
Could you add an integration test for custom class and controller names?

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go
Comment (eoinfennessy) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go]: Yes, makes sense. Thanks. Addressed with fixup to commit [0f5f150](https://github.com/maistra/istio/pull/1002/commits/0f5f15019aa7848b6cc4430bd3142ae44c873705).

File: tests/integration/servicemesh/maistra/maistra.go
Comment (eoinfennessy) [inline tests/integration/servicemesh/maistra/maistra.go:201]: It might be worth cleaning up this file after #983 merges. There are a number of functions such as `DisableWebhooksAndRestart`, `EnableIOR`, `DisableIOR` that duplicate the code in this function. These could be removed and the tests consuming them modified to use this function. Patch constants could also be moved to their consuming packages (unless used in multiple packages?).

Comment (jewertow) [reply tests/integration/servicemesh/maistra/maistra.go:201]: Let's wait for the IOR rebase. It will be merged very soon.

**Review by jewertow — COMMENTED** [review]
Great job! Let's wait for IOR and then we will merge it.

**Review by jewertow — APPROVED** [review]

**Pull request merged**
